libraries/AP_Scripting/applets/UniversalAutoLand.lua

```diff
@@ -0,0 +1,121 @@
+--]]
+
+--local variablees
+
+SCR_USER_EN = Parameter("SCR_USER1") -- enable this script
+local FIN_WP_DIST = param:get("SCR_USER2") -- distance of final wp from HOME
+local FIN_WP_ALT = param:get("SCR_USER3") --- altitude of final wp from HOME
+local TKOFF_ALT = param:get("TKOFF_ALT")
+local TKOFF_DIST = param:get("TKOFF_DIST")
+
+-- make sure you have a reasonable final angle if someone does not change defualts other than SCR1_USER to enable
+if FIN_WP_ALT < TKOFF_ALT then FIN_WP_ALT = TKOFF_ALT end
+if FIN_WP_DIST < TKOFF_DIST then FIN_WP_DIST = 2 * TKOFF_DIST end
+
+FRAME_GLOBAL = 3
+NAV_WAYPOINT = 16
+NAV_TAKEOFF = 22
+NAV_LAND = 21
+DO_LAND_START = 189
+
+function create_final_approach_WP(i,bearing,dist,alt) --index,degs,m,m
+   local item = mavlink_mission_item_int_t()
+   local loc = ahrs:get_home()
+   loc:offset_bearing(bearing,dist) ---degs and meters
+
+   item:seq(i)
+   item:frame(FRAME_GLOBAL)
+   item:command(NAV_WAYPOINT)
+   item:param1(0)
+   item:param2(0)
+   item:param3(0)
+   item:param4(0)
+   item:x(loc:lat())
+   item:y(loc:lng())
+   item:z(alt)
+   return item
+end
+
+function create_takeoff_WP()
+   local item = mavlink_mission_item_int_t()
+   local loc = ahrs:get_home()
+
+   item:seq(1)
+   item:frame(FRAME_GLOBAL)
+   item:command(NAV_TAKEOFF)
+   item:param1(15)
+   item:param2(0)
+   item:param3(0)
+   item:param4(0)
+   item:x(loc:lat())
+   item:y(loc:lng())
+   item:z(TKOFF_ALT)
+   return item
+end
+
+function create_land_WP()
+   local item = mavlink_mission_item_int_t()
+   local loc = ahrs:get_home()
+
+   item:seq(4)
+   item:frame(FRAME_GLOBAL)
+   item:command(NAV_LAND)
+   item:param1(15)
+   item:param2(0)
+   item:param3(0)
+   item:param4(0)
+   item:x(loc:lat())
+   item:y(loc:lng())
+   item:z(0)
+   return item
+end
+
+function create_do_land_start_WP()
+   local item = mavlink_mission_item_int_t()
+
+   item:seq(2)
+   item:frame(FRAME_GLOBAL)
+   item:command(DO_LAND_START)
+   item:param1(0)
+   item:param2(0)
+   item:param3(0)
+   item:param4(0)
+   item:x(0)
+   item:y(0)
+   item:z(0)
+   return item
+end
+
+function update()
+  if not arming:is_armed() then --if disarmed, wait until armed
+    mission_loaded = false
+    return update,1000
+  end
+  
+  if not mission_loaded then --if first time after arm and enabled is valid then create  mission 
+    local home = ahrs:get_home()
+    local location = ahrs:get_location()
+    if location:alt() > (home:alt() + param:get("TKOFF_LVL_ALT")*100) then
+        local yaw = ahrs:get_yaw()
+        mission:set_item(2,create_do_land_start_WP())
+        mission:set_item(3,create_final_approach_WP(3,math.deg(yaw)+180,FIN_WP_DIST,param:get("TKOFF_ALT")))
```

Comment:
You might want to wrap the yaw after adding 180 (e.g. 359 degrees + 180 = 539), Tridges scripted aerobatics has some wrap functions that will do it.

Reply:
had planned on it but forgot while debugging...done